### PR TITLE
(PC-33535) feat(offer): use mardown part in offer description

### PR DESCRIPTION
**Pull request by clesausse-pass**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-33535

## Flakiness

If I had to re-run tests in the CI due to flakiness, I add the incident [on Notion][2]

## Checklist

I have:

- [x] Made sure my feature is working on web.
- [x] Made sure my feature is working on mobile (depending on relevance : real or virtual devices)
- [x] Written **unit tests** native (and web when implementation is different) for my feature.
- [x] Added a **screenshot** for UI tickets or deleted the screenshot section if no UI change
- [ ] If my PR is a bugfix, I add the link of the "résolution de problème sur le bug" [on Notion][1]
- [x] I am aware of all the best practices and respected them.

## Screenshots

**delete** _if no UI change_

| Platform         | Mockup/Before | After |
| :--------------- | :-----------: | :---: |
| iOS              |Capture d’écran 2024-12-18 à 14 21 36|Capture d’écran 2024-12-18 à 14 21 06|
| Android          |Capture d’écran 2024-12-18 à 14 13 00|Capture d’écran 2024-12-18 à 14 12 46|
| Phone - Chrome   |Capture d’écran 2024-12-18 à 14 21 45|Capture d’écran 2024-12-18 à 14 20 17|
| Desktop - Chrome |Capture d’écran 2024-12-18 à 14 22 03|Capture d’écran 2024-12-18 à 14 19 52|

[1]: https://www.notion.so/passcultureapp/R-solution-de-probl-mes-sur-les-bugs-5dd6df8f6a754e6887066cf613467d0a
[2]: https://www.notion.so/passcultureapp/cb45383351b44723a6f2d9e1481ad6bb?v=10fe47258701423985aa7d25bb04cfee&pvs=4

## Best Practices

<details>
  <summary>Click to expand</summary>
These rules apply to files that you make changes to.
If you can't respect one of these rules, be sure to explain why with a comment.
If you consider correcting the issue is too time consuming/complex: create a ticket. Link the ticket in the code.

- In the production code: remove type assertions with `as` (type assertions are removed at compile-time, there is no runtime checking associated with a type assertion. There won’t be an exception or `null` generated if the type assertion is wrong). In certain cases `as const` is acceptable (for example when defining readonly arrays/objects). Using `as` in tests is tolerable.
- Remove bypass type checking with `any` (when you want to accept anything because you will be blindly passing it through without interacting with it, you can use `unknown`). Using `any` in tests is tolerable.
- Remove non-null assertion operators (just like other type assertions, this doesn’t change the runtime behavior of your code, so it’s important to only use `!` when you know that the value can’t be `null` or `undefined`).
- Remove all `@ts-expect-error` and `@eslint-disable`.
- Remove all warnings, and errors that we are used to ignore (`yarn test:lint`, `yarn test:types`, `yarn start:web`...).
- Use `gap` (`ViewGap`) instead of `<Spacer.Column />`, `<Spacer.Row />` or `<Spacer.Flex />`.
- Don't add new "alias hooks" (hooks created to group other hooks together). When adding new logic, this hook will progressively become more complex and harder to maintain.
- Remove logic from components that should be dumb.

Test specific:

- Avoid mocking internal parts of our code. Ideally, mock only external calls.
- When you see a local variable that is over-written in every test, mock it.
- Prefer `user` to `fireEvent`.
- When mocking feature flags, use `setFeatureFlags`. If not possible, mention which one(s) you want to mock in a comment (example: `jest.spyOn(useFeatureFlagAPI, 'useFeatureFlag').mockReturnValue(true) // WIP_NEW_OFFER_TILE in renderPassPlaylist.tsx` )
- In component tests, replace `await act(async () => {})` and `await waitFor(/* ... */)` by `await screen.findBySomething()`.
- In hooks tests, use `act` by default and `waitFor` as a last resort.
- Make a snapshot test for pages and modals ONLY.
- Make a web specific snapshot when your web page/modal is specific to the web.
- Make an a11y test for web pages.

Advice:

- Use TDD
- Use Storybook
- Use pair programming/mobs
</details>

### CODE REVIEW

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    expect(output).toEqual([
+      { text: 'Bold1', isBold: true },
+      { text: ' and ' },
+      { text: 'Bold2', isBold: true },
+      { text: '.' },
+    ])
+  })
+
+  it('should parse nested text with different styles', () => {
+    const input = '**Bold and _italic_**'
```
Comment (bebstein-pass) [inline src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:62]: il se passe quoi si je fais nawak ?

```md
en fermant les **tags _dans** le_ mauvais ordre

en ayant des __doubles_ _underscore__

est-ce que je peux \*\*échapper\*\* des \_characters\_ ?

ça fait quoi un tag vide **** ou __ ?

ou un tag vide mais pas vraiment, juste remplit avec un autre tag ****** ___

et si j'ouvre des tags sur plusieurs lignes **

et je le ferme quelques lignes plus loin**

et si les tags ne sont pas collés à des mots ** comme ça ** ou _ ceci _

ça fait quoi si j'ai une raw URL au milieu du texte ? https://perdu.com est-ce que j'ai toujours le **formattage** ?
```

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { MarkdownPartProps } from 'features/offer/types'
+
+// Regular expression to detect bold and italic
+const STYLES_REGEXP = /\*\*(.*?)\*\*|_(.*?)_/g
+
+export function parseMarkdown(markdown: string) {
+  function parse(markdown: string, styles: Record<string, boolean> = {}) {
```
Comment (bebstein-pass) [inline src/features/offer/helpers/parseMarkdown/parseMarkdown.ts:7]: pourquoi la fonction est définie dans l'autre fonction ?

File: src/ui/components/CollapsibleText/CollapsibleText.native.test.tsx [hunk 3]
```diff
@@ -30,100 +12,38 @@ jest.mock('libs/firebase/analytics/analytics')
...
+    )
 
-    fireEvent.press(screen.getByText('Voir plus'))
-
-    expect(screen.getByText(TEXT).props.numberOfLines).toBeUndefined()
+    expect(screen.getByTestId('styledBody')).toBeOnTheScreen()
   })
 
-  it('should use Réduire le texte in button accessibility label when pressing button', async () => {
+  it('should not use styled typo when text is not markdown', () => {
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleText.native.test.tsx:44]: j'ai l'impresion que ce test vérifie un détail d'implem

File: src/ui/components/MarkdownPart/MarkdownPart.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <StyledBody isUnderline={isUnderline} testID="styledBody">
+      {highlightLinks(text)}
+    </StyledBody>
+  )
+}
+
+const StyledBody = styled(TypoDS.Body)<{ isUnderline?: boolean }>(({ isUnderline }) => ({
+  textDecorationLine: isUnderline ? 'underline' : 'none',
+}))
```
Comment (bebstein-pass) [inline src/ui/components/MarkdownPart/MarkdownPart.tsx:43]: la fonction de style me semble etre répétée à l'identique, on pourrait l'extraire (j'ai pas d'idée de nom)

```suggestion
const underlineStyle = ({ isUnderline }) => ({
  textDecorationLine: isUnderline ? 'underline' : 'none',
})

const StyledBody = styled(TypoDS.Body)<{ isUnderline?: boolean }>(underlineStyle)
```

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import { ParsedDescription } from 'libs/parsers/highlightLinks'
+import { CollapsibleTextButton } from 'ui/components/CollapsibleText/CollapsibleTextButton/CollapsibleTextButton'
+import { TypoDS } from 'ui/theme'
+
+import { useIsTextEllipsis } from '../useIsTextEllipsis'
+
+type Props = {
+  expanded: boolean
+  numberOfLines: number
+  renderContent: () => React.JSX.Element[] | ParsedDescription
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.tsx:12]: pourquoi a-t-on besoin d'une fonction ? est-ce que des children seraient suffisant ?

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import { TypoDS } from 'ui/theme'
+
+type Props = {
+  expanded: boolean
+  numberOfLines: number
+  renderContent: () => React.JSX.Element[] | ParsedDescription
+  onButtonPress: VoidFunction
+}
+
+export const CollapsibleTextContent: FunctionComponent<Props> = ({
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx:16]: les `.ios.tsx` ne sont pas courant dans notre codebase

ça pourrait etre chouette d'avoir un commentaire qui explique la nécessité d'avoir un `.ios.tsx`

File: src/features/offer/components/OfferAbout/OfferAbout.tsx [hunk 7]
```diff
@@ -35,7 +35,7 @@ export const OfferAbout: FunctionComponent<Props> = ({
           {offer.description ? (
             <View>
               <TypoDS.BodyAccent>Description&nbsp;:</TypoDS.BodyAccent>
-              <CollapsibleText numberOfLines={NUMBER_OF_LINES_OF_DESCRIPTION_SECTION}>
+              <CollapsibleText numberOfLines={NUMBER_OF_LINES_OF_DESCRIPTION_SECTION} isMarkdown>
                 {offer.description}
               </CollapsibleText>
```
Comment (bebstein-pass) [inline src/features/offer/components/OfferAbout/OfferAbout.tsx:40]: je trouve ça bizarre de devoir ajouter une props assez spécifique : la gestion du markdown à un composant générique

je propose de garder le composant générique et d'avoir un composant qui s'occupe de transformer des children en markdown

```suggestion
              <CollapsibleText numberOfLines={NUMBER_OF_LINES_OF_DESCRIPTION_SECTION}>
                <Markdown>{offer.description}</Markdown>
              </CollapsibleText>
```

File: src/ui/components/CollapsibleText/CollapsibleText.tsx [hunk 8]
```diff
@@ -1,63 +1,44 @@
...
+    return highlightLinks(children)
+  }
 
   return (
-    <React.Fragment>
-      <TypoDS.Body
-        numberOfLines={expanded ? undefined : numberOfLines}
-        onLayout={onLayout}
-        onTextLayout={onTextLayout}>
-        {highlightLinks(children)}
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleText.tsx:39]: je crois que je suis responsable de ça, pardon

je trouve que `highlightLinks` n'a pas ça place dans le composant générique, il devrait plutot etre appélé sur les `children` là où on appelle `CollapsibleText`

**Review by bebstein-pass — COMMENTED** [review]
ça a du etre fun a faire, on a rarement l'occasion de faire des petits algos

---

j'aimerais que les plateformes jeunes et pro restent cohérentes dans le comportement du rendu

pour simplifier la maintenance, l'évolution, et la gestion des cas limites, je suis convaincu qu'il faudrait qu'on utilise une lib commune de parsing / render de markdown

il [existe plein de libs de markdown](https://www.npmjs.com/search?q=keywords:markdown), notamment

* [markdown-it](https://www.npmjs.com/package/markdown-it) mon préféré : écosystème plus riche, plus facilement customizable, plus facilement extensible
* [marked](https://www.npmjs.com/package/marked)

on en a déjà parlé, je sais que le ticket demande de le faire sans lib, je suis ok qu'on fasse une première itération sans lib, mais à la prochaine évolution sur le markdown, j'aimerais qu'on utilise une lib

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts (hunk 1 above)
Comment (clesausse-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:62]: ça le faisait pour tout sauf le tag vide ou le tag remplit d'un autre tag, bien vu ! j'ai amélioré

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.ts (hunk 2 above)
Comment (clesausse-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.ts:7]: C'était pas forcément utile, j'ai rectifié

File: src/ui/components/CollapsibleText/CollapsibleText.native.test.tsx (hunk 3 above)
Comment (clesausse-pass) [reply src/ui/components/CollapsibleText/CollapsibleText.native.test.tsx:44]: ça permet de mettre en avant que la MarkdownPart n'est pas utilisée. Tu veux que je supprime le test ?

File: src/ui/components/MarkdownPart/MarkdownPart.tsx (hunk 4 above)
Comment (clesausse-pass) [reply src/ui/components/MarkdownPart/MarkdownPart.tsx:43]: Du coup je l'ai carrément supprimée car on ne s'occupe pas de souligné pour le moment.

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts (hunk 1 above)
Comment (clesausse-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:62]: Capture d’écran 2024-12-19 à 16 55 30

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.tsx (hunk 5 above)
Comment (clesausse-pass) [reply src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.tsx:12]: C'est suffisant, j''ai update

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts (hunk 1 above)
Comment (bebstein-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:62]: sur ton screenshot, j'ai l'impresion qu'il manque une marge entre la fin du texte et le bouton "Voir moins" @mounouche-pass t'en penses quoi ?

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  if (!markdown) return parts
+
+  markdown.replace(STYLES_REGEXP, (match, isBold, isItalic, offset) => {
+    // Add plain text before the styled match
+    if (lastIndex < offset) {
+      parts.push({ text: markdown.slice(lastIndex, offset), ...styles })
+    }
+
+    // Determine the styles for the current match
+    const newStyles: Record<string, boolean> = { ...styles }
```
Comment (bebstein-pass) [inline src/features/offer/helpers/parseMarkdown/parseMarkdown.ts:22]: le type de styles est répété 2 fois

on pourrait l'extraire dans un type alias

File: src/features/offer/types.ts [hunk 10]
```diff
@@ -62,3 +62,9 @@ export type OfferContentProps = {
   searchGroupList: SearchGroupResponseModelv2[]
   subcategory: Subcategory
 }
+
+export type MarkdownPartProps = {
```
Comment (bebstein-pass) [inline src/features/offer/types.ts:66]: est-ce qu'on la partie markdown dans `offer` ?

est-ce qu'on le bouge dans `lib` ou `ui` ?

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import { CollapsibleTextButton } from 'ui/components/CollapsibleText/CollapsibleTextButton/CollapsibleTextButton'
+import { TypoDS } from 'ui/theme'
+
+type Props = {
+  expanded: boolean
+  numberOfLines: number
+  onButtonPress: VoidFunction
+} & PropsWithChildren
+
+// It is necessary to have ios.tsx file because the calculation to check that CollapsibleText has a button is different from web and android
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx:14]: qu'est-ce qu'on fait en web et android qui ne serait pas possible en ios ?

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.native.test.tsx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+const mockOnTextLayoutWhenTooShort = {
+  nativeEvent: {
+    lines: [{ width: 100 }, { width: 100 }, { width: 100 }, { width: 100 }, { width: 300 }],
+  },
+}
+
+const mockOnButtonPress = jest.fn()
+
+// Mock needed for the file test to take into account CollapsibleTextContent.tsx rather than CollapsibleTextContent.ios.tsx
+jest.mock('./CollapsibleTextContent', () => require('./CollapsibleTextContent.tsx'))
```
Comment (bebstein-pass) [inline src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.native.test.tsx:41]: est-ce que si on renomme en `CollapsibleTextContent.android.test.tsx` on peut faire sauter ce mock chelou ?

File: src/ui/components/Markdown/Markdown.tsx [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import React, { FunctionComponent } from 'react'
+
+import { parseMarkdown } from 'features/offer/helpers/parseMarkdown/parseMarkdown'
+import { MarkdownPartProps } from 'features/offer/types'
+import { MarkdownPart } from 'ui/components/MarkdownPart/MarkdownPart'
+
+type Props = {
+  text: string
```
Comment (bebstein-pass) [inline src/ui/components/Markdown/Markdown.tsx:8]: en react, généralement, quand on veut donner le contenu qui sera affiché dans un composant, on le passe entre les balises (`<Balise>le contenu</Balise>`), donc dans `children`

est-ce qu'il y a des raisons pour avoir choisi `text` et pas `children` ?

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+  it('should return an empty array when style text is empty', () => {
+    const input = '****'
+    const output = parseMarkdown(input)
+
+    expect(output).toEqual([])
+  })
+
+  it('should return an empty array when style text has an empty style text', () => {
+    const input = '********'
```
Comment (bebstein-pass) [inline src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:89]: d'autres cas (à la con) auxquels je viens de penser

```md
niveau 0 **niveau 1 _niveau 2 **niveau 3 (oui c'est débile) _niveau 4 (c'est encore plus débile)_**_** donc des trucs nested et en plus plein de closing tags qui s'enchainent

niveau 0 **niveau 1 _niveau 2 **niveau 3 _niveau 4_ ça devrait encore etre en italique car dicté par le niveau 2**_** est-ce que un tag fermant ferme le parent ?
```

**Review by bebstein-pass — APPROVED** [review]

File: src/features/offer/types.ts (hunk 10 above)
Comment (clesausse-pass) [reply src/features/offer/types.ts:66]: Les composants pour le Markdown sont dans `ui` mais pour l'instant ça sert uniquement à l'offre. Dans un souci de cohérence j'imagine c'est mieux de créer un fichier types dans ui

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx (hunk 11 above)
Comment (clesausse-pass) [reply src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.ios.tsx:14]: Tu peux jeter un oeil au fichier `usIsTextEllipsis.ios` supprimé (on se basait uniquement sur la dernière ligne mais avec une variation de style ça ne le faisait plus) puis `useIsTextEllipsis.android` et `useIsTextEllipsis`.
La solution initiale sur iOS ne fonctionnait plus en utilisant `parseMarkdown` donc on a fait en sorte d'afficher tout le texte puis de mettre le bouton et l'ellipsis si nécessaire. Cette logique fonctionnerait en android et web mais cela fait un blink qui se voit particulièrement en web. Sur iOS et Android ça ne se voit pas le temps du scroll. La solution n'étant pas optimum je ne l'ai pas reprise sur Android par contre j'ai créé un autre ticket pour essayer d'améliorer `CollapsibleText` car j'y ai passé beaucoup de temps déjà et ça sortait du périmètre de mon ticket d'origine.

File: src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.native.test.tsx (hunk 12 above)
Comment (clesausse-pass) [reply src/ui/components/CollapsibleText/CollapsibleTextContent/CollapsibleTextContent.native.test.tsx:41]: Et non, je l'ai tenté ^^ sans ça il me prend le fichier iOS. Je n'ai pas mieux à proposer.

File: src/ui/components/Markdown/Markdown.tsx (hunk 13 above)
Comment (clesausse-pass) [reply src/ui/components/Markdown/Markdown.tsx:8]: La fonctionne `parseMarkdown` s'attend à avoir un texte. Tu transformerais le `ReactNode` en `string` avant d'appeler la fonction ?

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts (hunk 14 above)
Comment (clesausse-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts:89]: tout ne se gère pas correctement mais je me demande si on veut rentrer dans ce niveau extrême de gestion étant donné que c'est n'imp ^^ et que ça ne plante pas l'app.

Capture d’écran 2024-12-23 à 10 26 40

File: src/ui/components/Markdown/Markdown.tsx (hunk 13 above)
Comment (bebstein-pass) [reply src/ui/components/Markdown/Markdown.tsx:8]: est-ce qu'on peut contraindre le `children` pour toujours avoir une `string` ?

```suggestion
  children: string
```

File: src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts (hunk 14 above)
Comment (bebstein-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts]: je ne sais pas ce qu'on veut faire

quand on passera à la lib, tout devrait etre mieux géré

Comment (clesausse-pass) [reply src/features/offer/helpers/parseMarkdown/parseMarkdown.test.ts]: Je viens d'en reparler à Henri, on a respecté ce qui a été demandé côté pro pour le moment et quand le temps sera venu d'en rediscuter, on en rediscutera :)

File: src/ui/components/Markdown/Markdown.tsx (hunk 13 above)
Comment (clesausse-pass) [reply src/ui/components/Markdown/Markdown.tsx]: Ah yes ça le fait ! je te laisse jeter un oeil au dernier commit si tu veux

File: src/ui/components/Markdown/Markdown.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+export const Markdown: FunctionComponent<Props> = ({ children }) => {
+  const parsedText: MarkdownPartProps[] = parseMarkdown(children)
+
+  return (
+    <React.Fragment>
+      {parsedText.map((part: MarkdownPartProps) => (
+        // A text can contain several times the same part therefore has no unique identifier
+        // If you have better than Math.random() you can update
+        <MarkdownPart key={`markdown-part-${Math.random()}`} {...part} />
```
Comment (mmeissonnier-pass) [inline src/ui/components/Markdown/Markdown.tsx:19]: Ici plutôt que d'utiliser Math.random() qui va forcer de rendre tous les morceaux à chaque rendu, pourquoi ne pas faire plutôt key={`markdown-part-${index)`} ? L'index venant des params de ton map. A priori l'ordre des tes morceaux de texte ne risque pas de changer.

Comment (clesausse-pass) [reply src/ui/components/Markdown/Markdown.tsx:19]: C'est ce que j'avais fait au début, dans tous les cas Sonar n'est pas content ^^
@bebstein-pass c'est bon pour toi si je remets `index` ?

**Review by mmeissonnier-pass — APPROVED** [review]

File: src/ui/components/Markdown/Markdown.tsx (hunk 15 above)
Comment (clesausse-pass) [reply src/ui/components/Markdown/Markdown.tsx:19]: @bebstein-pass après discussion avec Mathieu, je vais remettre l'index je pense que dans le cas présent ça n'est pas gênant car il n'y aura pas de réordonnancements et donc pas de re-rendu

Comment (bebstein-pass) [reply src/ui/components/Markdown/Markdown.tsx:19]: ça me va

**Review by bebstein-pass — APPROVED** [review]

**Pull request merged**